### PR TITLE
[DRAFT] Column Defaults

### PR DESCRIPTION
**Pull request by Hydrocharged**
The most rough draft thing ever. This PR seems FAR larger than it actually is. `information_schema.go` and `parse_test.go` make up like 4500 lines of change, but I just put those large maps inside of a function, which tabbed all of the lines over by one, but results in a massive diff.

I ran into a ton of import cycles while trying to make this all work. This PR does not have any cycles, but it's far from an elegant solution. I feel the true solution is to compartmentalize the `sql` package (and probably other packages while we're at it), but that would be a **ton** of work for integrators (and probably this project too, idk how good the refactoring tools are for moving between packages). So this is what I've got in the meantime.

The overall flow with this implementation would be for an integrator to store the expression string that's returned from `CREATE/ALTER TABLE`, which can be fetched by just calling `String()` on the returned expression. When an integrator is loading their table's schemas into the engine, they'd call `sql.StringToColumnDefaultValue`, which would give them the expression back. In the event that an integrator only wants to grab a default value, then they don't have to instantiate the entire engine, and can just call the aforementioned function and call `Eval()` on the expression.

I'm also considering moving the parsing code out of the `parse` package and into the default functions, which would be duplicating quite a bit of code (specifically `exprToExpression`), but would remove the testing dependency on the `parse` package.

### CODE REVIEW

File: sql/columndefault.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+func MustStringToColumnDefaultValue(ctx *Context, exprStr string) *ColumnDefaultValue {
+	expr, err := StringToColumnDefaultValue(ctx, exprStr)
+	if err != nil {
+		panic(err)
+	}
+	return expr
+}
+
+// ColumnDefaultValue is an expression representing the default value of a column. May represent both a default literal
+// and a default expression. A nil pointer is a valid instance of this struct, and all method calls will return without error.
```
Comment (zachmu) [inline sql/columndefault.go:50]: Not true right now :p

File: sql/columndefault.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+func NewColumnDefaultValue(expr Expression, representsLiteral bool) *ColumnDefaultValue {
+	return &ColumnDefaultValue{
+		Expression: expr,
+		literal:    representsLiteral,
+	}
+}
+
+// Children implements sql.Expression
+func (e *ColumnDefaultValue) Children() []Expression {
+	return []Expression{e.Expression}
```
Comment (zachmu) [inline sql/columndefault.go:68]: Need a nil check so you can return a zero-length expression slice

File: sql/expression/columndefault.go [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+	"fmt"
+	"github.com/liquidata-inc/go-mysql-server/sql"
+)
+
+// This function is to prevent import cycles. Since the sql package cannot reference the expression package, we have to
+// push the required function to a global that will be referenced from within the sql package. Otherwise, we'd need to
+// duplicate a lot of code or split up sql into smaller packages. If a better way is found, then feel free to replace this.
+func init() {
+	sql.ColumnDefaultManager.ExpressionTransformUp = func(inputExpr sql.Expression) (sql.Expression, error) {
+		return TransformUp(inputExpr, func(e sql.Expression) (sql.Expression, error) {
```
Comment (zachmu) [inline sql/expression/columndefault.go:13]: Don't do this. If you need to resolve columns, functions, etc. the place to do so is in the analyzer, on an insert plan node that would use a default. If you want to just do validation (make sure every expression type is allowed), then use sql.Inspect.

File: sql/columndefault.go [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+		panic(err)
+	}
+	return expr
+}
+
+// ColumnDefaultValue is an expression representing the default value of a column. May represent both a default literal
+// and a default expression. A nil pointer is a valid instance of this struct, and all method calls will return without error.
+type ColumnDefaultValue struct {
+	Expression
+	literal bool
```
Comment (zachmu) [inline sql/columndefault.go:53]: Not sure this type is pulling its weight. Why is the distinction of being a literal important or not enough to break out a new type? Why isn't a columns' default a sql.Expression, which can be nil? What value does this type add?

And if it is a new type, it's a new type of expression and should live in the expressions package.

File: sql/columndefault.go [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+	AllowedDefaultFunctions FunctionRegistry
+	ExpressionTransformUp func(Expression) (Expression, error)
+	ParseDefaultString func(*Context, string) (Expression, error)
+}
+var ColumnDefaultManager = &columnDefaultManager{
+	AllowedDefaultFunctions: NewFunctionRegistry(),
+}
+
+// StringToColumnDefaultValue takes in a string representing a default value and returns the equivalent Expression.
+func StringToColumnDefaultValue(ctx *Context, exprStr string) (*ColumnDefaultValue, error) {
```
Comment (zachmu) [inline sql/columndefault.go:18]: Why doesn't this need to live here, rather than in the parse package?

File: sql/columndefault.go (hunk 4 above)
Comment (Hydrocharged) [reply sql/columndefault.go:53]: I had this initially in the `expressions` package, before getting an import cycle and moved it here. The position of everything was because I ran into import cycles. The `expressions` package references the `sql` package (which this lives in), so I would be unable to directly reference this type from within `sql`.

So the literal distinction does have importance to it. `DEFAULT 2` and `DEFAULT (2)` have a different set of rules that they abide by, and that includes functions such as `NOW()` and `(NOW())` (the former is viewed under the _literal_ rules). In a nutshell, all valid expressions can be applied to all types, however literals can only apply to specific types (which depends on the literal). For example, `col INT UNSIGNED DEFAULT -1` is invalid, however `col INT UNSIGNED DEFAULT (-1)` is valid, even though it will cause an error whenever the column is not specified.

A different example would be `col VARCHAR(20) DEFAULT NOW()` vs `col VARCHAR(20) DEFAULT (NOW())`. The former will error, however the latter will not only succeed at creation, but correctly convert the date into a string on insertion.

Additionally, the integrators would only be storing a string (we could make it a string and boolean), so we want that string to parse back into the same thing (a literal or expression). When stringifying this, we can include an outer set of parentheses if it's not a literal.

The alternative to this implementation would be an expression and a boolean, but I feel they should be in the same structure since they're tied together. I'm definitely treating this more as a struct that implements `sql.Expression` rather than an expression in the `expressions` package, since this is specific to column defaults.

File: sql/expression/columndefault.go (hunk 3 above)
Comment (Hydrocharged) [reply sql/expression/columndefault.go:13]: That works when you're on the standard SQL path, but if the idea is to have integrators store the default expression as a string, how would they be able to fetch the default outside of the SQL path, such as `dolt`'s import path? The current implementation is geared so that integrators can keep their non-SQL write paths while accessing SQL's concept of default values. For things like column values being referenced outside of the SQL path, I was imagining a callback of some sort that the integrator may provide for each column. This would just be a function call on the `*ColumnDefaultValue` type. If we changed the default value to a standard `sql.Expression`, then we could include the same on the `*Column` type itself.

File: sql/columndefault.go (hunk 5 above)
Comment (Hydrocharged) [reply sql/columndefault.go:18]: I'm keeping it by the `ColumnDefaultValue` type, as it's the primary way to instantiate it as far as integrators are concerned. It appears that the `sql` package is the place for most of the interfaces and types for integrators to add resides, so this follows that trend.

I remember running into an import cycle with having all of this done in the `parse` package before, but I've rewritten quite a lot since then, so perhaps it might not happen again. I'll check.

**Comment by Hydrocharged** [comment]
After reading some of your comments, there's another approach that I haven't given a spin yet. I don't feel that I'll truly love any solution with our current package structure, but I'll give it a try and see how it works.

File: sql/columndefault.go (hunk 5 above)
Comment (Hydrocharged) [reply sql/columndefault.go]: This has been moved to the `parse` package in the latest iteration. Still not a fan of the instantiation of a type not being right beside the type.

**Comment by Hydrocharged** [comment]
So I've restructured the changes a bit, moved a few things around to prevent import cycles (like `information_schema.go` getting its own package), and made a few changes to visibility of some methods. You'll like this far more than the previous implementation.

Of note, `*ColumnDefaultValue` is still in the `sql` package. I experimented with moving it to the `expression` package once more, replacing it with a dedicated interface in the `sql` package, but doing so essentially negates its nil-value usefulness. I find `nil` to be an elegant way to represent the difference between an explicitly-declared default value and implicitly-declared one, so I like that both forms allow for method calls. I still feel like the default value should not be viewed as an expression though, and rather as a discrete type that implements the `Expression` interface.

File: sql/columndefault.go (hunk 4 above)
Comment (zachmu) [reply sql/columndefault.go:12]: I didn't understand the distinction between an expression and a literal in this context. That makes sense.

Still belongs in the expression package though. Seems like the main drawback to putting it there is that you wouldn't be able to declare it as an *expression.ColumnDefault in the sql package (because of circular imports), which is not ideal, but not a deal breaker. What do we lose by the sql package considering the field to be a generic Expression rather than this specific one?

File: sql/parse/parse.go [hunk 6]
```diff
@@ -1354,6 +1358,75 @@ func selectExprsToExpressions(ctx *sql.Context, se sqlparser.SelectExprs) ([]sql
...
+	// The literal and expression distinction seems to be decided by the presence of parentheses, even for defaults like NOW() vs (NOW())
+	// 2+2 would evaluate to a literal under the parentheses check, but will have children due to being an Arithmetic expression, thus we check for children.
+	return ExpressionToColumnDefaultValue(ctx, parsedExpr, len(parsedExpr.Children()) == 0 && !strings.HasPrefix(exprStr, "("))
+}
+
+// ExpressionToColumnDefaultValue takes in an Expression and returns the equivalent ColumnDefaultValue if the expression
+// is valid for a default value. If the expression represents a literal (and not an expression that returns a literal, so "5"
+// rather than "(5)"), then the parameter "isLiteral" should be true.
+func ExpressionToColumnDefaultValue(ctx *sql.Context, inputExpr sql.Expression, isLiteral bool) (*sql.ColumnDefaultValue, error) {
+	expr, err := expression.TransformUp(inputExpr, func(e sql.Expression) (sql.Expression, error) {
```
Comment (zachmu) [inline sql/parse/parse.go:1393]: There needs to be a really, really good reason to do this here instead of at analyzer time. Expressions aren't expected to be eval-able at parse time, which is what we're doing here.

File: sql/parse/parse.go [hunk 7]
```diff
@@ -1354,6 +1358,75 @@ func selectExprsToExpressions(ctx *sql.Context, se sqlparser.SelectExprs) ([]sql
...
+			if err != nil {
+				return nil, err
+			}
+			resolvedFunc, err := builtInFunc.Call(expr.Arguments...)
+			if err != nil {
+				return nil, err
+			}
+			return resolvedFunc, nil
+		case *expression.UnresolvedColumn:
+			//TODO: handle this
```
Comment (zachmu) [inline sql/parse/parse.go:1407]: I mean... the way you handle this is passing it through the analyzer to be resolved

File: sql/parse/parse.go [hunk 8]
```diff
@@ -1354,6 +1358,75 @@ func selectExprsToExpressions(ctx *sql.Context, se sqlparser.SelectExprs) ([]sql
...
+	})
+	if err != nil {
+		return nil, err
+	}
+	//TODO: currently (2+2)/2 will, when output as a string, give (2 + 2 / 2), which is clearly wrong
+	return sql.NewColumnDefaultValue(expr, isLiteral), nil
+}
+
+// MustStringToColumnDefaultValue is StringToColumnDefaultValue except that it panics on errors.
+func MustStringToColumnDefaultValue(ctx *sql.Context, exprStr string) *sql.ColumnDefaultValue {
```
Comment (zachmu) [inline sql/parse/parse.go:1422]: Is this test only? If so, put it in the test files (and maybe test packages)(

File: sql/parse/parse_test.go [hunk 9]
```diff
@@ -26,849 +26,793 @@ var showCollationProjection = plan.NewProject([]sql.Expression{
...
-	),
-	"DESCRIBE FORMAT=tree SELECT * FROM foo": plan.NewDescribeQuery(
-		"tree", plan.NewProject(
-			[]sql.Expression{expression.NewStar()},
-			plan.NewUnresolvedTable("foo", ""),
-		)),
-	"DESC FORMAT=tree SELECT * FROM foo": plan.NewDescribeQuery(
-		"tree", plan.NewProject(
-			[]sql.Expression{expression.NewStar()},
+func init() {
```
Comment (zachmu) [inline sql/parse/parse_test.go:31]: This change makes the diff impossible to review. Can you summarize why this can no longer be statically initialized?

File: sql/parse/parse.go [hunk 10]
```diff
@@ -53,6 +53,13 @@ var (
 )
 
 var describeSupportedFormats = []string{"tree"}
+var allBuiltInFuncs = sql.NewFunctionRegistry()
+
+func init() {
+	if err := allBuiltInFuncs.Register(function.Defaults...); err != nil {
```
Comment (zachmu) [inline sql/parse/parse.go:59]: Again, this shouldn't be needed.

**Comment by Hydrocharged** [comment]
> I didn't understand the distinction between an expression and a literal in this context. That makes sense.
> 
> Still belongs in the expression package though. Seems like the main drawback to putting it there is that you wouldn't be able to declare it as an *expression.ColumnDefault in the sql package (because of circular imports), which is not ideal, but not a deal breaker. What do we lose by the sql package considering the field to be a generic Expression rather than this specific one?

I can't comment on it for some reason, so I'll just reply here. So, at least from an integrator's standpoint, I'm not wanting default values to be something that are only available when used on the SQL path. An integrator should be able to have their own write path, and yet also able to make use of the default outside of that. With the default value being it's own discrete type, there will be additionally methods to the type to allow its usage outside of just being a standard expression. This is all toward usability outside of the SQL engine.

At the moment, the only additional information that's stored within is whether the value is a literal or expression. I'm planning on extending that, so just an expression and boolean will not cover those additional methods. That's the reasoning for making it a discrete type. And with it being a discrete type, it cannot live in the `expression` package due to import cycles.

File: sql/parse/parse.go (hunk 8 above)
Comment (Hydrocharged) [reply sql/parse/parse.go:1422]: It is not test only, any internal tables that contain default values that are created on engine creation make use of this.

File: sql/parse/parse.go (hunk 6 above)
Comment (Hydrocharged) [reply sql/parse/parse.go:1393]: I'll address all of the analyzer-related comments here. The reason that I am not doing this is that it ties any integrator that wants to make use of default values to the SQL write path, unless we create a separate entry point in `go-mysql-server` that's specifically for default values.

As an example, in the `dolt` project we have an import path for CSV files, and it currently does not make use of anything from the SQL path. With the way that we're planning on implementing default values (where the integrator stores a string), the only way for them to make use of the default value is with a reliance on the SQL engine. This implementation is geared so that the reliance is on a function call, rather than forcing all write paths through the SQL engine (which I don't find a reasonable solution).

We could make a special entrypoint (mentioned earlier), so that it's `engine.CreateDefaultValue(str)`, as `engine.Query(str)` is the only one that currently exists. The other alternative would be to serialize expressions, so that the integrator can store a byte array, which may be the easier option between the two. But there may be another option that I'm not thinking of.

File: sql/columndefault.go (hunk 4 above)
Comment (Hydrocharged) [reply sql/columndefault.go:12]: So, at least from an integrator's standpoint, I'm not wanting default values to be something that are only available when used on the SQL path. An integrator should be able to have their own write path, and yet also able to make use of the default outside of that. With the default value being it's own discrete type, there will be additionally methods to the type to allow its usage outside of just being a standard expression. This is all toward usability outside of the SQL engine.

At the moment, the only additional information that's stored within is whether the value is a literal or expression. I'm planning on extending that, so just an expression and boolean will not cover those additional methods. That's the reasoning for making it a discrete type. And with it being a discrete type, it cannot live in the expression package due to import cycles.

File: sql/parse/parse_test.go (hunk 9 above)
Comment (Hydrocharged) [reply sql/parse/parse_test.go]: This is a relic from when I was using `init()` functions everywhere. This has been fixed.

**Comment by Hydrocharged** [comment]
I'll open another PR with the changes discussed earlier when that's ready. This one has a lot of comments on it that will soon be irrelevant.
